torch/_inductor/simple_fsdp.py

```diff
@@ -0,0 +1,154 @@
+# mypy: allow-untyped-defs
+# pyre-strict
+import typing
+from typing import (
+    Any,
+    Counter,
+    DefaultDict,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+)
+from enum import Enum
+from . import config, ir, comms, scheduler
+from collections import OrderedDict
+
+import torch
+
+from .dependencies import WeakDep
+from .utils import is_collective, is_wait, tuple_sorted
+
+import torch.distributed._functional_collectives
+_c10d_functional = torch.ops._c10d_functional
+
+class NodeType(Enum):
+    ALL_GATHER = 0
+    WAIT = 1
+    COMP = 2
+    REDUCE_SCATTER = 3
+    COVERT_ELEMENT = 4
+
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_before_last_wait: Optional[bool] = True,
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_before_last_wait set to True: all_gather_i is reordered before wait_i-1
+    2. all_gather_before_last_wait set to False: all_gather_i is reordered after wait_i-1
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
```

Comment:
consider renaming it to something like `results` to differentiate it from other intermediate lists.